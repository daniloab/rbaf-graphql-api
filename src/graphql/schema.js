import { GraphQLSchema } from 'graphql'

import QueryType from './types/QueryType'
import MutationType from './types/MutationType'

export const schema = new GraphQLSchema({
    query: QueryType,
    mutation: MutationType
})