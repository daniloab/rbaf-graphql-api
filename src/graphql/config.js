import path from 'path'
import dotenv from 'dotenv'

const root = path.join.bind(this, __dirname, '../')

const ENV = process.env

// Database Settings
const dBdevelopment = ENV.MONGO_URL || 'mongodb://localhost/database'
const dBproduction = ENV.MONGO_URL || 'mongodb://localhost/database'

// Test Database Settings
// const test = 'mongodb://localhost/awesome-test';

// Export DB Settings
export const databaseConfig = ENV.NODE_ENV === 'production' ? dBproduction : dBdevelopment;

// Export GraphQL Server settings
export const graphqlPort = ENV.GRAPHQL_PORT || 9001;
export const jwtSecret = ENV.JWT_KEY || 'jwt_secret';