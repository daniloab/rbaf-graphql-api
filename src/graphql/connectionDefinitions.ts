import {
  GraphQLBoolean,
  GraphQLFieldConfigArgumentMap,
  GraphQLFieldConfigMap,
  GraphQLFieldResolver,
  GraphQLInt,
  GraphQLList,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLString,
  Thunk,
} from "graphql";

export const forwardConnectionArgs: GraphQLFieldConfigArgumentMap = {
  after: {
    type: GraphQLString,
  },
  first: {
    type: GraphQLInt,
  },
};

export const backwardConnectionArgs: GraphQLFieldConfigArgumentMap = {
  before: {
    type: GraphQLString,
  },
  last: {
    type: GraphQLInt,
  },
};

export const connectionArgs: GraphQLFieldConfigArgumentMap = {
  ...forwardConnectionArgs,
  ...backwardConnectionArgs,
};

type ConnectionConfig = {
  name?: string | null;
  nodeType: GraphQLObjectType;
  resolveNode?: GraphQLFieldResolver<any, any> | null;
  resolveCursor?: GraphQLFieldResolver<any, any> | null;
  edgeFields?: Thunk<GraphQLFieldConfigMap<any, any>> | null;
  connectionFields?: Thunk<GraphQLFieldConfigMap<any, any>> | null;
};

export type GraphQLConnectionDefinitions = {
  edgeType: GraphQLObjectType;
  connectionType: GraphQLObjectType;
};

const pageInfoType = new GraphQLObjectType({
  name: "PageInfoExtended",
  description: "Information about pagination in a connection.",
  fields: () => ({
    hasNextPage: {
      type: GraphQLNonNull(GraphQLBoolean),
      description: "When paginating forwards, are there more items?",
    },
    hasPreviousPage: {
      type: GraphQLNonNull(GraphQLBoolean),
      description: "When paginating backwards, are there more items?",
    },
    startCursor: {
      type: GraphQLString,
      description: "When paginating backwards, the cursor to continue.",
    },
    endCursor: {
      type: GraphQLString,
      description: "When paginating forwards, the cursor to continue.",
    },
  }),
});

function resolveMaybeThunk<T>(thingOrThunk: Thunk<T>): T {
  return typeof thingOrThunk === "function"
    ? (thingOrThunk as () => T)()
    : thingOrThunk;
}

export function connectionDefinitions(
  config: ConnectionConfig
): GraphQLConnectionDefinitions {
  const { nodeType, resolveCursor, resolveNode } = config;
  const name = config.name || nodeType.name;
  const edgeFields = config.edgeFields || {};
  const connectionFields = config.connectionFields || {};

  const edgeType = new GraphQLObjectType({
    name: `${name}Edge`,
    description: "An edge in a connection.",
    fields: () => ({
      node: {
        type: nodeType,
        resolve: resolveNode,
        description: "The item at the end of the edge",
      },
      cursor: {
        type: GraphQLNonNull(GraphQLString),
        resolve: resolveCursor,
        description: "A cursor for use in pagination",
      },
      ...(resolveMaybeThunk(edgeFields) as any),
    }),
  });

  const connectionType = new GraphQLObjectType({
    name: `${name}Connection`,
    description: "A connection to a list of items.",
    fields: () => ({
      count: {
        type: GraphQLInt,
        description: "Number of items in this connection",
      },
      totalCount: {
        type: GraphQLInt,
        resolve: (connection) => connection.count,
        description: `A count of the total number of objects in this connection, ignoring pagination.
  This allows a client to fetch the first five objects by passing "5" as the
  argument to "first", then fetch the total count so it could display "5 of 83",
  for example.`,
      },
      startCursorOffset: {
        type: GraphQLNonNull(GraphQLInt),
        description: "Offset from start",
      },
      endCursorOffset: {
        type: GraphQLNonNull(GraphQLInt),
        description: "Offset till end",
      },
      pageInfo: {
        type: GraphQLNonNull(pageInfoType),
        description: "Information to aid in pagination.",
      },
      edges: {
        type: GraphQLNonNull(GraphQLList(edgeType)),
        description: "A list of edges.",
      },
      ...(resolveMaybeThunk(connectionFields) as any),
    }),
  });

  return { edgeType, connectionType };
}
