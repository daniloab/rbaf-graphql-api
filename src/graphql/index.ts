export { debugConsole } from "./debugConsole";
export { createLoader } from "./createLoader";
export { withConnectionCursor } from "./withConnectionCursor";
export { registerLoader, getDataloaders } from "./loaderRegister";
export { connectionArgs, connectionDefinitions } from "./connectionDefinitions";
