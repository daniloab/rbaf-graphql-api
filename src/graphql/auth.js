import jwt from 'jsonwebtoken'
import { User } from './model'
import { jwtSecret } from './config'

export async function getUser (token) {
  if (!token) return { user: null }

  try {
    const decodedToken = jwt.verify(token.substring(4), jwtSecret)

    const user = await User.findOne({ _id: decodedToken.id })

    return {
      user,
    }
  } catch (err) {
    return { user: null }
  }
}

export function generateToken (user) {
  return `JWT ${jwt.sign({ id: user._id }, jwtSecret)}`
}