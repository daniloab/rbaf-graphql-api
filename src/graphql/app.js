require('dotenv').config()
import Koa from 'koa'
import GraphQLHTTP from 'koa-graphql'
import Router from 'koa-router'
import cors from '@koa/cors'
import bodyParser from 'koa-bodyparser'

import { getUser } from './auth';
import { schema } from './schema'

const app = new Koa()
const router = new Router();

const graphqlSettingsPerReq = async req => {
    const { user } = await getUser(req.header.authorization)

    return {
        graphiql: process.env.NODE_ENV !== 'production',
        schema,
        context: {
            user,
            req,
        },
        formatError: error => {
            console.log(error.message);
            console.log(error.locations);
            console.log(error.stack);

            return {
                message: error.message,
                locations: error.locations,
                stack: error.stack,
            };
        },
    }
}

const graphqlServer = GraphQLHTTP(graphqlSettingsPerReq)

router.all('/graphql', graphqlServer)

app.use(bodyParser())
app.use(cors())
app.use(router.routes()).use(router.allowedMethods());

export default app
// app.use(mount('/graphql', GraphQLHTTP({
//     schema: graphQLSchema,
//     rootValue: graphQLResolvers,
//     graphiql: true
// })));

// app.listen(GRAPHQL_PORT, () => console.log(
//     `GraphQL Server is now running on http://localhost:${GRAPHQL_PORT}/graphql`
// ))

// app.on('error', err => {
//     log.error('server error', err)
// })