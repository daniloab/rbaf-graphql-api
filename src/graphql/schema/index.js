import { buildSchema } from 'graphql'

export default buildSchema(`
    type Team {
        _id: ID!
        name: String!
        img: String!
        players: [Player!]
    }

    type Player {
        _id: ID!
        name: String!
    }

    input TeamInput {        
        name: String!
        img: String!
    }

    input PlayerInput {
        name: String!
    }


    type Mutation {
        createTeam(teamInput: TeamInput): Team
        createPlayer(playerInput: PlayerInput): Player
    }
    
    type Query {
        teams: [Team!]!
        players: [Player!]!
    }
`)