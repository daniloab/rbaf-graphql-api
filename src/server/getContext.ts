import { Request } from "koa";

import { IUser } from "./modules/user/UserModel";
import { getDataloaders } from "../graphql/loaderRegister";
// import { GraphQLContext } from "./graphql/types";

type GraphQLContext = {};

type ContextVars = {
  user?: IUser | null;
  req?: Request;
};

export const getContext = async (ctx: ContextVars) => {
  const dataloaders = getDataloaders();

  return {
    req: ctx.req,
    dataloaders,
    user: ctx.user,
  } as GraphQLContext;
};
