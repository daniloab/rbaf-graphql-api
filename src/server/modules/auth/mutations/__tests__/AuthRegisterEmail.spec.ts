import { graphql } from "graphql";
import {
  clearDbAndRestartCounters,
  connectMongoose,
  disconnectMongoose,
  sanitizeTestObject,
} from "../../../../../../test";

import { createUser } from "../../../user/fixture/createUser";
import { getContext } from "../../../../getContext";

import { schema } from "../../../../schema";
import { createTeam } from "../../../team/fixture/createTeam";

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

it("should create a new user", async () => {
  const team = await createTeam({
    name: "Team A",
  });

  const query = `
    mutation M ($name: String!, $email: String!, $password: String!, ) {
      UserRegisterEmailByEmail (input: { name: $name, email: $email, password: $password }) {
        error
        token
      }
    }
  `;
  const variables = {
    name: "Danilo",
    email: "danilo@test.com",
    password: "awesomepw",
  };

  const rootValue = {};

  const context = await getContext({ team });

  const result = await graphql(schema, query, rootValue, context, variables);

  expect(result.errors).toBeUndefined();
  expect(result.data.UserRegisterEmailByEmail.error).toBeNull();

  expect(result.data.UserRegisterEmailByEmail.token).toBeDefined();

  expect(sanitizeTestObject(result.data, ["token"])).toMatchSnapshot();
});
