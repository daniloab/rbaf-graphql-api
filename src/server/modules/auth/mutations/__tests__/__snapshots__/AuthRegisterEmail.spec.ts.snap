// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a new user 1`] = `
Object {
  "UserRegisterEmailByEmail": Object {
    "error": "EMPTY",
    "token": "FROZEN-TOKEN",
  },
}
`;
